services/galley/test/integration/API.hs

```diff
@@ -4477,4 +4487,225 @@ testDefederationNotifications = do
 
+-- Testing defederation notifications. The important thing to note for all
+-- of this is that when defederating from a remote domain only _2_ notifications
+-- are sent, and both are identical. One notification is at the start of
+-- defederation, and one is sent at the end of defederation. No other
+-- notifications about users being removed from conversations, or conversations
+-- being deleted are sent. We are do not want to DOS either our local clients,
+-- nor our own services.
+-- There are four tests here.
+
+-- * A normal run where we have users from both remote domains in a conversation. Both remote users should be removed.
+
+-- * A no-op run where we have no remote users in the conversation. The conversation remains unchanged.
+
+-- * A domain A biased run where we have a conversation with a remote member from domain A, but none from domain B. The conversation remains unchanged.
+
+-- * A domain B biased run where we have a conversation with a remote member from domain B, but none from domain A. The conversation remains unchanged.
+
+testConnectionRemovedNotifications :: TestM ()
+testConnectionRemovedNotifications = do
+  -- alice, bob are in a team
+  (tid, alice, [bob]) <- createBindingTeamWithQualifiedMembers 2
+
+  -- charlie is a local guest
+  charlie <- randomQualifiedUser
+  connectUsers (qUnqualified alice) (pure (qUnqualified charlie))
+
+  let remoteDomain1 = Domain "far-away.example.com"
+      remoteDomain2 = Domain "far-away-2.example.com"
+  -- dee and erin are remote guests
+  dee <- Qualified <$> randomId <*> pure remoteDomain1
+  erin <- Qualified <$> randomId <*> pure remoteDomain2
+
+  connectWithRemoteUser (qUnqualified alice) dee
+  connectWithRemoteUser (qUnqualified alice) erin
+
+  -- they are all in a local conversation
+  conv <-
+    responseJsonError
+      =<< postConvWithRemoteUsers
+        (qUnqualified alice)
+        Nothing
+        defNewProteusConv
+          { newConvQualifiedUsers = [bob, charlie, dee, erin],
+            newConvTeam = Just (ConvTeamInfo tid)
+          }
+        <!! const 201 === statusCode
+
+  c <- view tsCannon
+  WS.bracketRN c (map qUnqualified $ [alice, bob, charlie, dee, erin]) $ \[wsA, wsB, wsC, wsD, wsE] -> do
+    -- conversation access role changes to team only
+    (_, reqs) <- withTempMockFederator' (mockReply ()) $ do
+      -- Remove the connection
+      connectionRemovedFederation remoteDomain1 remoteDomain2 !!! const 200 === statusCode
+      -- First notification to local clients
+      WS.assertMatchN_ (5 # Second) ([wsA, wsB, wsC]) $
+        wsAssertFederationConnectionRemoved remoteDomain1 remoteDomain2
+      -- Second notification to local clients
+      WS.assertMatchN_ (5 # Second) ([wsA, wsB, wsC]) $
+        wsAssertFederationConnectionRemoved remoteDomain1 remoteDomain2
+      -- dee and erin's remotes don't receive a notification
+      WS.assertNoEvent (5 # Second) [wsD, wsE]
```

Comment:
Should we also assert that no event as a result of the cleanup is sent to the remaining users?

Reply:
Done